### PR TITLE
html representation for jupyter notebooks

### PR DESCRIPTION
**Pull request by percygautam**
## Description
This PR implements the html representation for arviz.InferenceData object in jupyter notebooks. Related to discussion in #1066.

## Checklist

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/master/CONTRIBUTING.md#pull-request-checklist) PR format
- [x] Code style  correct (follows pylint and black guidelines)
- [ ] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/master/CHANGELOG.md#v0xx-unreleased)

### CODE REVIEW

File: arviz/data/inference_data.py [hunk 1]
```diff
@@ -133,6 +133,18 @@ def __repr__(self):
             msg += "\n\nWarmup iterations saved ({}*).".format(WARMUP_TAG)
         return msg
 
+    def _repr_html_(self):
+        """Make html representation of object"""
```
Comment (canyon289) [inline arviz/data/inference_data.py:137]: Would change object to `InferenceData` so this docstring can be read without having to refer to class name

File: arviz/data/inference_data.py [hunk 2]
```diff
@@ -133,6 +133,18 @@ def __repr__(self):
...
+        xr.set_options(display_style="html")
+        html_template, element_template, css_template = html_templates(self)
+        template, elements = "", ""
+        for data_group in self._groups:
+            func = getattr(self, data_group)
+            elements += element_template.format(data_group, func._repr_html_())
+        template += html_template.format(elements)
```
Comment (canyon289) [inline arviz/data/inference_data.py:144]: Are these `+=` operators adding strings together? If so could this be changed to another method of string concat? The reason I ask is that `+=` is typically used for numbers or other things and it's confusing to see it used as strings.

Its typically not used for string concats for the reasons detailed in the article below

https://wiki.python.org/moin/PythonSpeed/PerformanceTips#String_Concatenation

File: arviz/utils.py [hunk 3]
```diff
@@ -664,3 +665,35 @@ def credible_interval_warning(credible_interval, hdi_prob):
 
     hdi_prob = credible_interval
     return hdi_prob
+
+
+def html_templates():
```
Comment (canyon289) [inline arviz/utils.py:670]: I would change this python function to a class that of structure. Reason for this is that

1. Function doesn't do anything but return values
2. Defining objects in a function scope makes it hard to "get at them" without running function
3. In future use case we only want one of these things we still need to deal with the other 2 because with a function call we always get them
4. In the code where this is used it'll make it more clear that computation likely isn't happening but were just "grabbing a template"



```
HTMLTemplates:
    html_template = xyz
```

As a sidenote DataClasses would be great but they were introduced in 3.7 so if we use them here they will break ArviZ for Python 3.6 users

File: arviz/data/inference_data.py (hunk 2 above)
Comment (OriolAbril) [reply arviz/data/inference_data.py:144]: I would also rename `func` to `ds`, `group` or `data` so the name is related to the object it stores. Maybe something like:

```
for group in self._groups_all:
    ds = getattr(...
```

I'd also use `_groups_all` given that groups start hidden and users can expand the desired ones

File: arviz/data/inference_data.py [hunk 4]
```diff
@@ -133,6 +133,18 @@ def __repr__(self):
             msg += "\n\nWarmup iterations saved ({}*).".format(WARMUP_TAG)
         return msg
 
+    def _repr_html_(self):
+        """Make html representation of object"""
+        xr.set_options(display_style="html")
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:138]: I am not sure about this. html repr is the default for xarray now (we could require a version where this is already default) so this will generally have no effect, and if it were not the default, it means the user has specifically changed it so undoing the change in the background and with no option to override it seems like a bad idea.

How does xarray choose between html repr and "regular" print? Maybe we can check xarray defaults us too to choose one or the other?

File: arviz/data/inference_data.py [hunk 5]
```diff
@@ -133,6 +133,18 @@ def __repr__(self):
...
+        html_template, element_template, css_template = html_templates(self)
+        template, elements = "", ""
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:140]: I'd rename the `template` string to something more suited, maybe `html_repr`?

File: arviz/utils.py [hunk 6]
```diff
@@ -664,3 +665,35 @@ def credible_interval_warning(credible_interval, hdi_prob):
...
+              <div class='xr-header'>
+                <div class="xr-obj-type">arviz.InferenceData</div>
+              </div>
+              {}
+            </div>
+            """
+    element_template = """
+            <ul class="xr-sections">
+            <li class = "xr-section-item">
+                  <input id="xr-dataset-coord_{0}" class="xr-section-summary-in" type="checkbox">
```
Comment (OriolAbril) [inline arviz/utils.py:682]: given that the group name is added to the `id` I am guessing the id is arbitrary and the class is what defines the format used. Is that so? If so, can this be renamed so it becomes more clear what is happening? 

If I am not mistaken, in `{1}` we put the whole html_repr of the xarray dataset corresponding to the group at hand, so having xr, dataset and coord but no az, idata nor group in id section is confusing.

File: arviz/data/inference_data.py (hunk 2 above)
Comment (percygautam) [reply arviz/data/inference_data.py:144]: @canyon289 Thanks for the suggestion. I'd keep it in mind for the future. I also learnt a lot of other tricks too from the article :)

File: arviz/utils.py (hunk 3 above)
Comment (percygautam) [reply arviz/utils.py:670]: Okay, I understand. I've changed it to class.

File: arviz/data/inference_data.py (hunk 4 above)
Comment (percygautam) [reply arviz/data/inference_data.py:138]: I understand the issue, I tried to solve it by saving the default style and reapplying after the display of html_repr.
xarray is doing like [this](https://github.com/jsignell/xarray/blob/1d960933ab252e0f79f7e050e6c9261d55568057/xarray/core/common.py#L138) The default option is "text".

File: arviz/utils.py (hunk 6 above)
Comment (percygautam) [reply arviz/utils.py:682]: Yes, the id is arbitrary. I left it as it is when I first tinkered with xarray's code and forgot to change it. I've changed it.

**Comment by percygautam** [comment]
@canyon289 Yes the xarray's repr is in html if we set the option. This PR is making the whole inferenceData object dynamic. The result can be seen in this jupyter notebook: inhttps://nbviewer.jupyter.org/github/percygautam/arviz-examples/blob/master/HTML%20repr.ipynb

File: arviz/data/inference_data.py (hunk 4 above)
Comment (OriolAbril) [reply arviz/data/inference_data.py:141]: I would mimic xarray behaviour. If display style is text we should not print the html repr. 

The default in latest xarray is already html https://github.com/pydata/xarray/blob/master/xarray/core/options.py#L23 so we should probably require xarray>=0.15.1. I think we should have html repr by default so requiring this or later xarray version would make it default.

File: arviz/utils.py [hunk 7]
```diff
@@ -664,3 +665,34 @@ def credible_interval_warning(credible_interval, hdi_prob):
...
+    html_template = """
+            <div>
+              <div class='xr-header'>
+                <div class="xr-obj-type">arviz.InferenceData</div>
+              </div>
+              {}
+            </div>
+            """
+    element_template = """
+            <ul class="xr-sections">
```
Comment (OriolAbril) [inline arviz/utils.py:680]: Currently the vertical spacing between groups is very large. I think this could be due to `ul` being part of the element template as this is defining a different list for each group. Can you try moving it to the html template so that we have a single list (only one `ul` defined in html template) and the groups are then elements of this single list (one `li` per group)

File: arviz/data/inference_data.py (hunk 4 above)
Comment (percygautam) [reply arviz/data/inference_data.py:141]: Okay, I understand. I was reading the fork of the guy who implemented the html_repr for xarray a long time ago.

**Comment by OriolAbril** [comment]
LGTM, we can merge after updating a couple notebooks.

**Comment by canyon289** [comment]
I am going to request tests before merge. IMO all code should be tested if it can, and this is no exception :)

File: arviz/data/inference_data.py [hunk 8]
```diff
@@ -125,14 +126,38 @@ def __init__(self, **kwargs):
...
         msg = "Inference data with groups:\n\t> {options}".format(
             options="\n\t> ".join(self._groups)
         )
         if self._groups_warmup:
             msg += "\n\nWarmup iterations saved ({}*).".format(WARMUP_TAG)
         return msg
 
+    def _repr_html_(self):
+        """Make html representation of InferenceData object."""
+        from xarray.core.options import OPTIONS
```
Comment (canyon289) [inline arviz/data/inference_data.py:139]: Why is this import here? Shouldnt it be at the top of the module?

File: arviz/utils.py [hunk 9]
```diff
@@ -664,3 +665,32 @@ def credible_interval_warning(credible_interval, hdi_prob):
 
     hdi_prob = credible_interval
     return hdi_prob
+
+
+class HtmlTemplate:
+    """Contains html templates."""
```
Comment (canyon289) [inline arviz/utils.py:671]: Expand docstring. Contains html templates for what?

Would change this to something along the lines of contains html templates for InferenceData repr

File: arviz/utils.py [hunk 10]
```diff
@@ -664,3 +665,32 @@ def credible_interval_warning(credible_interval, hdi_prob):
...
+                <div class="xr-obj-type">arviz.InferenceData</div>
+              </div>
+              <ul class="xr-sections">
+              {}
+              </ul>
+            </div>
+            """
+    element_template = """
+            <li class = "xr-section-item">
+                  <input id="idata_{0}" class="xr-section-summary-in" type="checkbox">
```
Comment (canyon289) [inline arviz/utils.py:685]: Instread of using numbers for string interpolation, using named arguments would make this easier to read and harder to mess up

File: arviz/utils.py [hunk 11]
```diff
@@ -664,3 +665,32 @@ def credible_interval_warning(credible_interval, hdi_prob):
...
+                  <label for="idata_{0}" class = "xr-section-summary">{0}</label>
+                  <div class="xr-section-inline-details"></div>
+                  <div class="xr-section-details">
+                      <ul id="xr-dataset-coord-list" class="xr-var-list">
+                          <div style="padding-left:2rem;">{1}<br></div>
+                      </ul>
+                  </div>
+            </li>
+            """
+    specific_style = ".xr-wrap{width:700px!important;}"
```
Comment (canyon289) [inline arviz/utils.py:695]: No change required but just curious what is so specific about this style?

**Comment by canyon289** [comment]
@percygautam this is already looking great and youre on the right track. Hope you dont mind my additional comments :)

File: arviz/utils.py (hunk 11 above)
Comment (percygautam) [reply arviz/utils.py:695]: The xarray uses the style `min-width:400px; max-width:700px;` for sections. For single xarray dataset max-width is used, while when used in sub-section min-width is used.
So, we are overwriting our own style over xarray's style.

**Comment by percygautam** [comment]
> I am going to request tests before merge. IMO all code should be tested if it can, and this is no exception :). I would add a test that calls `_repr_html_` directly and tests that the html string is returned.
> 
> I would also add a tests the correct repr is returned without an explicit call. I assume this will require some mocking and to "trick" the non jupyter notebook test to provide the html repr.

I have added the basic test. How do I add the second test? I looked through but couldn't find the way.

File: arviz/data/inference_data.py (hunk 4 above)
Comment (OriolAbril) [reply arviz/data/inference_data.py:141]: could it be you modified that and rolled back the changes? I have a vague memory of `f"<pre>{escape(repr(self))}</pre>"` but now I an not sure. Either way, this clearly has to be changed to use the xarray default instead of overriding it

File: arviz/utils.py [hunk 12]
```diff
@@ -664,3 +665,36 @@ def credible_interval_warning(credible_interval, hdi_prob):
 
     hdi_prob = credible_interval
     return hdi_prob
+
+
+class HtmlTemplate:
+    """Contains html templates for InferenceData repr."""
```
Comment (OriolAbril) [inline arviz/utils.py:671]: I think you'll need this to keep both black and numpydoc happy: https://github.com/arviz-devs/arviz/blob/master/arviz/rcparams.py#L28

File: arviz/utils.py (hunk 7 above)
Comment (OriolAbril) [reply arviz/utils.py:682]: This `ul`/`li` change is also marked as resolved but it is not and again I have a vague memory of seeing the `ul` moved to html instead of element template :/

**Comment by percygautam** [comment]
Actually I did revert some commits as I have forgot to gitignore the cache of notebooks and pushed them. I'll check and make all the changes again.

File: arviz/data/inference_data.py (hunk 4 above)
Comment (percygautam) [reply arviz/data/inference_data.py]: Yes, that's the case. I apologize for these discrepancies. I have corrected the ones I found.

File: arviz/tests/base_tests/test_data.py [hunk 13]
```diff
@@ -456,6 +456,13 @@ def test_map(self, use):
         )
         assert np.allclose(idata_map.posterior.mu, idata.posterior.mu)
 
+    def test_repr_html(self):
+        idata = load_arviz_data("centered_eight")
```
Comment (canyon289) [inline arviz/tests/base_tests/test_data.py:460]: NIT, test should have docstring

File: arviz/tests/base_tests/test_data.py [hunk 14]
```diff
@@ -456,6 +456,13 @@ def test_map(self, use):
...
+        html = idata._repr_html_()  # pylint: disable=protected-access
+
+        assert html is not None
```
Comment (canyon289) [inline arviz/tests/base_tests/test_data.py:463]: This is gerat

File: arviz/utils.py (hunk 11 above)
Comment (canyon289) [reply arviz/utils.py:695]: Cool!

File: arviz/data/inference_data.py [hunk 15]
```diff
@@ -125,14 +127,38 @@ def __init__(self, **kwargs):
...
         msg = "Inference data with groups:\n\t> {options}".format(
             options="\n\t> ".join(self._groups)
         )
         if self._groups_warmup:
             msg += "\n\nWarmup iterations saved ({}*).".format(WARMUP_TAG)
         return msg
 
+    def _repr_html_(self):
+        """Make html representation of InferenceData object."""
+        display_style = OPTIONS["display_style"]
```
Comment (canyon289) [inline arviz/data/inference_data.py:140]: > > I am going to request tests before merge. IMO all code should be tested if it can, and this is no exception :). I would add a test that calls `_repr_html_` directly and tests that the html string is returned.
> > I would also add a tests the correct repr is returned without an explicit call. I assume this will require some mocking and to "trick" the non jupyter notebook test to provide the html repr.
> 
> I have added the basic test. How do I add the second test? I looked through but couldn't find the way.

This is going to be  fun one. Making tests is sometimes a puzzle and harder than the actual programming itself, but important if you want you feature to be stable and not broken by any others.

In the test you'll have to monkeypatch `OPTIONS` to so when the test executes the value is `text` to test that this if condition works. Just being up front here it's going to be hard and confusing, but youre going to learn a lot more about python.

I don't want to block this whole PR for that one thing though. We can merge but I really would like you to add this test in a followup PR. At a minimum create an issue so we can track


https://docs.pytest.org/en/latest/monkeypatch.html

Comment (OriolAbril) [reply arviz/data/inference_data.py:141]: I think it will not be necessary to monekypatch because the display style option is already customizable similarly to an rcParam. We can start the test by setting the `display_style` to `text`/`html`, make sure the correct repr is returned and then finish the test by going back to the original displaystyle (much like your original implementation of the repr @percygautam).

File: arviz/tests/base_tests/test_data.py [hunk 16]
```diff
@@ -456,6 +456,13 @@ def test_map(self, use):
...
+        assert "<div" in html
```
Comment (OriolAbril) [inline arviz/tests/base_tests/test_data.py:464]: I would make this assert more specific to check all required elements are present (at least partially). i.e. for html repr to work, we need our custom part (we could check for group names being present for example), the dataset repr should be present too (we could check for variable names of for Coordinate section title being present) and also style inherited from xarray (we could test a generic line/command from the style to assert is is correclty being appended to the repr)

File: arviz/data/inference_data.py (hunk 15 above)
Comment (canyon289) [reply arviz/data/inference_data.py:141]: Its typically bad practice to have tests rely on global state like rcparam although in this case lets just do it cause monkeypatching would be annoying :D

Comment (percygautam) [reply arviz/data/inference_data.py:141]: I am okay with both options. We can currently implement the test using rcParams, and create the issue to track the monkeypatch test and merge this PR. I will implement the test later.
The `html_repr` really facilitates to check other methods that I am adding, so it'll be better to merge this one as soon as possible.

Comment (canyon289) [reply arviz/data/inference_data.py:141]: Agree with merging as soon as possible! Seems like youre only a couple linting errors away :)

**Comment by canyon289** [comment]
Add changelog and this one should be good to go

**Comment by OriolAbril** [comment]
@percygautam can you rebase and rerun notebook? `from_pyjags` has just been added so it is one more case for which to show the new repr
